### PR TITLE
feat(core-snapshots-cli): initial implementation

### PR DESCRIPTION
**Pull request by kristjank**
## Proposed changes
The purpose of this plugin is to provide local snapshot functionality, so in case of issues the blockchain can be rebuild locally from own exported data.
The plugin provides a cli interface, with the following available commands:

- create (with option to append, or to select `--start` and `--end` block height to export)
- import
- verify
- rollback (and backup forged transactions during the process)

Please read the readme for instructions: https://github.com/ArkEcosystem/core/blob/feat/snapshosts-new/packages/core-snapshots-cli/README.md

## Todo (remaining)
- [x] split into two plugins (cli and core-snapshots) #1187 
- [x] add ark codecs to support serialisation
- [x] add ark-lite codecs to support faster serialisation
- [x] use custom db (its own, or provided by constructor) 
- [x] archive and compress older/appended snapfile 
- [x] truncate bugy
- [x] run migrations on fresh pg

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read the [CONTRIBUTING](https://docs.ark.io/developers/guidelines/contributing.html) documentation
- [x] I have added necessary documentation (if appropriate)

## Further comments
Tests will follow in another PR.

### CODE REVIEW

File: packages/core-snapshots-cli/lib/transport/index.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+'use strict'
+const fs = require('fs-extra')
+const QueryStream = require('pg-query-stream')
+const JSONStream = require('JSONStream')
+const msgpack = require('msgpack-lite')
+const delay = require('delay')
+
+// const stream = require('spex')(Promise).stream
```
Comment (supaiku0) [inline packages/core-snapshots-cli/lib/transport/index.js]: Can be removed?

File: packages/core-snapshots-cli/lib/db/queries/blocks/height-range.sql [hunk 2]
```diff
@@ -0,0 +1,20 @@
+SELECT
+  ID,
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/blocks/height-range.sql]: Properties in queries should be all lower-case, only SQL keywords should be uppercase.

File: packages/core-snapshots-cli/lib/db/queries/blocks/latest.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+SELECT *
+FROM BLOCKS
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/blocks/latest.sql]: Properties in queries should be all lower-case, only SQL keywords should be uppercase.

File: packages/core-snapshots-cli/lib/db/queries/index.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    deleteFromHeight: loadQueryFile(__dirname, './blocks/delete-from-height.sql')
+  },
+  transactions: {
+    timestampRange: loadQueryFile(__dirname, './transactions/timestamp-range.sql'),
+    deleteFromTimestamp: loadQueryFile(__dirname, './transactions/delete-from-timestamp.sql')
+  },
+  rounds: {
+    deleteFromRound: loadQueryFile(__dirname, './rounds/delete-from-round.sql')
+  },
+  transactionsExport: (start, end) => {
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/index.js]: All SQL queries should be in files and be loaded via `loadQueryFile` to make sure we get all benefits of https://vitaly-t.github.io/pg-promise/QueryFile.html.

File: packages/core-snapshots-cli/lib/db/queries/index.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    return `SELECT id, block_id, version, sequence, timestamp, sender_public_key, recipient_id, type, vendor_field_hex, amount, fee, serialized from TRANSACTIONS WHERE TIMESTAMP BETWEEN ${start} AND ${end} ORDER BY TIMESTAMP`
+  },
+  blocksExport: (start, end) => {
+    return `SELECT id, version, timestamp, previous_block, height, number_of_transactions, total_amount, total_fee, reward, payload_length, payload_hash, generator_public_key, block_signature FROM BLOCKS WHERE HEIGHT BETWEEN ${start} AND ${end} ORDER BY HEIGHT`
+  },
+  transactionsBackup: (start) => {
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/index.js]: All SQL queries should be in files and be loaded via `loadQueryFile` to make sure we get all benefits of https://vitaly-t.github.io/pg-promise/QueryFile.html.

File: packages/core-snapshots-cli/lib/db/queries/index.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    timestampRange: loadQueryFile(__dirname, './transactions/timestamp-range.sql'),
+    deleteFromTimestamp: loadQueryFile(__dirname, './transactions/delete-from-timestamp.sql')
+  },
+  rounds: {
+    deleteFromRound: loadQueryFile(__dirname, './rounds/delete-from-round.sql')
+  },
+  transactionsExport: (start, end) => {
+    return `SELECT id, block_id, version, sequence, timestamp, sender_public_key, recipient_id, type, vendor_field_hex, amount, fee, serialized from TRANSACTIONS WHERE TIMESTAMP BETWEEN ${start} AND ${end} ORDER BY TIMESTAMP`
+  },
+  blocksExport: (start, end) => {
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/index.js]: All SQL queries should be in files and be loaded via `loadQueryFile` to make sure we get all benefits of https://vitaly-t.github.io/pg-promise/QueryFile.html.

File: packages/core-snapshots-cli/lib/db/queries/index.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    return `SELECT id, version, timestamp, previous_block, height, number_of_transactions, total_amount, total_fee, reward, payload_length, payload_hash, generator_public_key, block_signature FROM BLOCKS WHERE HEIGHT BETWEEN ${start} AND ${end} ORDER BY HEIGHT`
+  },
+  transactionsBackup: (start) => {
+    return `SELECT id, sequence, serialized from TRANSACTIONS WHERE TIMESTAMP > ${start}`
+  },
+  truncateTable: (table) => {
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/index.js]: All SQL queries should be in files and be loaded via `loadQueryFile` to make sure we get all benefits of https://vitaly-t.github.io/pg-promise/QueryFile.html.

File: packages/core-snapshots-cli/lib/db/queries/rounds/delete-from-round.sql [hunk 8]
```diff
@@ -0,0 +1,2 @@
+DELETE FROM ROUNDS
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/rounds/delete-from-round.sql]: Properties in queries should be all lower-case, only SQL keywords should be uppercase.

File: packages/core-snapshots-cli/lib/db/queries/transactions/delete-from-timestamp.sql [hunk 9]
```diff
@@ -0,0 +1,2 @@
+DELETE FROM TRANSACTIONS
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/transactions/delete-from-timestamp.sql]: Properties in queries should be all lower-case, only SQL keywords should be uppercase.

File: packages/core-snapshots-cli/lib/db/queries/transactions/timestamp-range.sql [hunk 10]
```diff
@@ -0,0 +1,19 @@
+SELECT
+  ID,
```
Comment (faustbrian) [inline packages/core-snapshots-cli/lib/db/queries/transactions/timestamp-range.sql]: Properties in queries should be all lower-case, only SQL keywords should be uppercase.

File: packages/core-snapshots-cli/lib/db/queries/index.js (hunk 4 above)
Comment (kristjank) [reply packages/core-snapshots-cli/lib/db/queries/index.js]: I agree, wanted to move all queries there, but they don't work with streams, at least I couldn't find a way of getting parsed query to raw text from loadQuery format, as it is parsed on the pg-server. The querystream uses custom pg-native implementation and that is why only string is accepted there as it forwards it to the cursor.

File: packages/core-snapshots-cli/lib/db/queries/index.js (hunk 5 above)
Comment (kristjank) [reply packages/core-snapshots-cli/lib/db/queries/index.js]: I agree, wanted to move all queries there, but they don't work with streams, at least I couldn't find a way of getting parsed query to raw text from loadQuery format, as it is parsed on the pg-server. The querystream uses custom pg-native implementation and that is why only string is accepted there as it forwards it to the cursor.

File: packages/core-snapshots-cli/lib/manager.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
+'use strict'
+const container = require('@arkecosystem/core-container')
+const logger = container.resolvePlugin('logger')
+const Database = require('./db')
+const utils = require('./utils')
+const { exportTable, importTable, verifyTable, backupTransactionsToJSON } = require('./transport')
+
+module.exports = class SnapshotManager {
```
[the remark this answers is not in the document]
Comment (kristjank) [reply packages/core-snapshots-cli/lib/manager.js]: done

File: packages/core-snapshots-cli/lib/commands/rollback.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+const logger = container.resolvePlugin('logger')
+const SnapshotManager = require('../manager')
+
+module.exports = async (options) => {
+  if (options.height === -1) {
+    logger.warn('Rollback height is not specified. Rolling back to last completed round.')
+  }
+  logger.info(`Starting the process of blockchain rollback to block height of ${options.height}`)
+
+  await new SnapshotManager().rollbackChain(options)
```
[the remark this answers is not in the document]
Comment (kristjank) [reply packages/core-snapshots-cli/lib/commands/rollback.js]: done

File: packages/core-snapshots-cli/lib/commands/verify.js [hunk 13]
```diff
@@ -0,0 +1,6 @@
+'use strict'
+const SnapshotManager = require('../manager')
+
+module.exports = async (options) => {
+  await new SnapshotManager().verifyData(options)
```
[the remark this answers is not in the document]
Comment (kristjank) [reply packages/core-snapshots-cli/lib/commands/verify.js]: done

File: packages/core-snapshots-cli/lib/commands/create.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+const SnapshotManager = require('../manager')
+const utils = require('../utils')
+
+module.exports = async (options) => {
+  if (options.filename && !fs.existsSync(utils.getPath(options.filename))) {
+    logger.error(`Appending not possible. Existing snapshot ${options.filename} not found. Exiting...`)
+    process.exit(1)
+  }
+
+  await new SnapshotManager().exportData(options)
```
[the remark this answers is not in the document]
Comment (kristjank) [reply packages/core-snapshots-cli/lib/commands/create.js]: done

**Comment by kristjank** [comment]
closes #238

File: packages/core-snapshots-cli/lib/db/queries/index.js (hunk 7 above)
Comment (kristjank) [reply packages/core-snapshots-cli/lib/db/queries/index.js]: I agree, wanted to move all queries there, but they don't work with streams, at least I couldn't find a way of getting parsed query to convert to raw text from loadQuery format, as it is parsed on the pg-server. The querystream uses custom pg-native implementation and that is why only string is accepted there as it forwards it to the cursor. Its a compromise

**Comment by faustbrian** [comment]
Is there any way to programmatically invoke this? To do things like proposed in https://github.com/ArkEcosystem/core/issues/1151 we would need to be able to call all methods from within core itself and the CLI aspect should be a secondary thought.

**Comment by kristjank** [comment]
yes @faustbrian good catch. the cli part can be simply extracted our, as for usage from programmatically view point the Snapshot manager should do it (with a change to pass database connection as a parameter - if used in core, the rest is functional).

Do you want to split this now or after we close this one?

**Comment by faustbrian** [comment]
I would split it now so we can be sure they work without any issues before we move to mainnet where we would to avoid any issues with snapshots.

**Comment by fix** [comment]
general comments:
- msgpack is impressively fast
- looking at raw data, there is a lot redundancy (the properties are repeated for each data). Not sure if possible, but if we can reduce this as well we could divide by 2 the size of the data stores (and likely the speed)

**Comment by fix** [comment]
in that respect in order to improve space storage (and thus i believe speed), what about having 2 msgpack codec: BlockCodec and TransactionCodec?

https://github.com/kawanet/msgpack-lite#custom-extension-types-codecs

**Comment by faustbrian** [comment]
Looks like the merge got messed up because now 234 files have been modified in this PR which makes it impossible to properly review it.

File: packages/core-snapshots/lib/db/index.js [hunk 15]
```diff
@@ -71,7 +76,7 @@ module.exports = class Database {
 
     if (!startBlock || !endBlock) {
       logger.error('Wrong input height parameters for building export queries. Blocks at height not found in db.')
-      process.exit(1)
+      throw new Error('Wrong input height parameters for building export queries. Blocks at height not found in db.')
```
Comment (faustbrian) [inline packages/core-snapshots/lib/db/index.js]: This should be a `process.exit(1)` to 100% make sure the application actually exits or set `process. exitCode = 1` before you throw the error.

File: packages/core-snapshots/lib/db/index.js [hunk 16]
```diff
@@ -8,20 +8,25 @@ const { rawQuery } = require('./utils')
...
     try {
       const pgp = require('pg-promise')({ promiseLib: promise })
       this.pgp = pgp
       this.db = pgp(container.resolveOptions('database').connection)
       this.__createColumnSets()
+      logger.info('Snapshots: Database connected')
     } catch (error) {
       logger.error(`Error while connecting to postgres: ${error}`)
-      process.exit(1)
+      throw new Error(error.stack)
```
Comment (faustbrian) [inline packages/core-snapshots/lib/db/index.js]: This should be a `process.exit(1)` to 100% make sure the application actually exits or set `process. exitCode = 1` before you throw the error.

File: packages/core-database-postgres/lib/queries/spv/multi-signatures.sql [hunk 17]
```diff
@@ -2,4 +2,4 @@ SELECT sender_public_key,
        serialized
 FROM transactions
 WHERE TYPE = 4
-ORDER BY created_at DESC
+ORDER BY (timestamp + sequence) DESC
```
[the remark this answers is not in the document]
Comment (kristjank) [reply packages/core-database-postgres/lib/queries/spv/multi-signatures.sql]: sequence is the order of transactions included in the blocks (similar to mempool sequence)
timestamp is block timestamp

i think the created at can be removed now yes

File: packages/core-snapshots-cli/README.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
+![ARK Core](https://i.imgur.com/1aP6F2o.png)
```
[the remark this answers is not in the document]
Comment (kristjank) [reply packages/core-snapshots-cli/README.md]: ?

**Pull request merged**